packages/vm/src/imports.rs

```diff
@@ -90,28 +90,63 @@ pub fn do_remove<S: Storage, Q: Querier>(ctx: &mut Ctx, key_ptr: u32) -> VmResul
     Ok(())
 }
 
-pub fn do_canonicalize_address<A: Api>(
+pub fn do_canonicalize_address<A: Api, S: Storage, Q: Querier>(
     api: A,
     ctx: &mut Ctx,
     source_ptr: u32,
     destination_ptr: u32,
-) -> VmResult<()> {
-    let human: HumanAddr = read_string_region(ctx, source_ptr, MAX_LENGTH_HUMAN_ADDRESS)?.into();
-    let canon = api.canonical_address(&human)?;
-    write_region(ctx, destination_ptr, canon.as_slice())?;
-    Ok(())
+) -> VmResult<u32> {
+    let source_data = read_region(ctx, source_ptr, MAX_LENGTH_HUMAN_ADDRESS)?;
+    if source_data.is_empty() {
+        return Ok(create_error_region::<S, Q>(ctx, "Input is empty")?);
+    }
+
+    let source_string = match String::from_utf8(source_data) {
+        Ok(s) => s,
+        Err(_) => {
+            return Ok(create_error_region::<S, Q>(
+                ctx,
+                "Input is not valid UTF-8",
+            )?)
+        }
+    };
+    let human: HumanAddr = source_string.into();
+    match api.canonical_address(&human) {
+        Ok(canon) => {
+            write_region(ctx, destination_ptr, canon.as_slice())?;
+            Ok(0)
+        }
+        // This check is unrelyable since we cannot tell by the error type if the error should be
+        // reported to the contract or indicates a broken backend.
+        // Err(FfiError::Other { error, .. }) => Ok(create_error_region::<S, Q>(ctx, &error)?),
+        Err(error) => Err(error.into()),
+    }
 }
 
 pub fn do_humanize_address<A: Api>(
     api: A,
     ctx: &mut Ctx,
     source_ptr: u32,
     destination_ptr: u32,
-) -> VmResult<()> {
+) -> VmResult<u32> {
     let canonical = Binary(read_region(ctx, source_ptr, MAX_LENGTH_CANONICAL_ADDRESS)?);
+    // TODO: how to report API errors back to the contract?
     let human = api.human_address(&CanonicalAddr(canonical))?;
     write_region(ctx, destination_ptr, human.as_str().as_bytes())?;
-    Ok(())
+    Ok(0)
+}
+
+fn create_error_region<S: Storage, Q: Querier>(ctx: &mut Ctx, string: &str) -> VmResult<u32> {
+    let error_ptr = with_func_from_context::<S, Q, u32, u32, _, _>(ctx, "allocate", |allocate| {
+        let out_size = to_u32(string.len())?;
+        let ptr = allocate.call(out_size)?;
+        if ptr == 0 {
+            return Err(CommunicationError::zero_address().into());
+        }
+        Ok(ptr)
+    })?;
+    write_region(ctx, error_ptr, string.as_bytes())?;
+    Ok(error_ptr)
```

Comment:
This function isn't really specific to errors. I'd rename it to something like `write_to_contract_memory`, change the `string: &str` parameter to a `buffer: &[u8]` parameter, and change the places that call `with_func_from_context::<S, Q, u32, u32, _, _>(ctx, "allocate", ...)` to call this function.

Reply:
Very good idea!

Reply:
Implemented in 2673a5e and 7553541